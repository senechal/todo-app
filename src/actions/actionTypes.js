export const ADD_ITEM = 'TODO//ADD_ITEM';
export const REMOVE_ITEM = 'TODO//REMOVE_ITEM';
export const CHANGE_ITEM = 'TODO//CHANGE_ITEM';
export const TOGGLE_DONE = 'TODO//TOGGLE_DONE';
export const TOOGLE_ALL_DONE = 'TODO//TOOGLE_ALL_DONE';
export const START_EDITING = 'TODO//START_EDITING';
export const END_EDITING = 'TODO//END_EDITING';
