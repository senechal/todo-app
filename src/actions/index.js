import {
  ADD_ITEM,
  CHANGE_ITEM,
  REMOVE_ITEM,
  TOGGLE_DONE,
  TOOGLE_ALL_DONE,
  START_EDITING,
  END_EDITING,
} from './actionTypes';

export const addItem = () => ({ type: ADD_ITEM });

export const changeItem = (value, id) => ({
  type: CHANGE_ITEM,
  value,
  id,
});

export const removeItem = (id) => ({
  type: REMOVE_ITEM,
  id,
});

export const toggleItem = (id) => ({ type: TOGGLE_DONE, id });

export const toggleAll = () => ({ type: TOOGLE_ALL_DONE });

export const startEditing = (id) => ({ type: START_EDITING, id });

export const endEditing = () => ({ type: END_EDITING });
