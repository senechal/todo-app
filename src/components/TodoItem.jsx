import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import {
  changeItem, removeItem, toggleItem, startEditing, endEditing,
} from 'actions';
import { TEXT_DELETE_ITERM, TEST_ID_TODO_ITEM_CHECKBOX, TEST_ID_TODO_ITEM_DELETE_BUTTON } from 'utils/constants';
import {
  Item, Checkbox, DeleteButton, DeleteIcon,
} from './styles';

import InlineEditInput from './InlineEditInput';

const TodoItem = ({ value, id, done }) => {
  const dispatch = useDispatch();

  const handleInputChange = useCallback((event) => {
    const { target: { value: eventValue } } = event;
    dispatch(changeItem(eventValue, id));
  }, [dispatch, id]);

  const handleRemove = useCallback(() => {
    dispatch(removeItem(id));
  }, [dispatch, id]);

  const handleToggleItem = useCallback(() => {
    dispatch(toggleItem(id));
  }, [dispatch, id]);

  const handleStartEdit = useCallback(() => {
    dispatch(startEditing(id));
  }, [dispatch, id]);

  const handleEndEdit = useCallback(() => {
    dispatch(endEditing());
    if (value === '') {
      dispatch(removeItem(id));
    }
  }, [dispatch, id, value]);

  return (
    <Item>
      <Checkbox data-testid={TEST_ID_TODO_ITEM_CHECKBOX} checked={done} onChange={handleToggleItem} />
      <InlineEditInput
        value={value}
        onChange={handleInputChange}
        onStartEdit={handleStartEdit}
        onEndEdit={handleEndEdit}
      />
      <DeleteButton
        data-testid={TEST_ID_TODO_ITEM_DELETE_BUTTON}
        title={TEXT_DELETE_ITERM}
        onClick={handleRemove}
      >
        <DeleteIcon />
      </DeleteButton>
    </Item>
  );
};

TodoItem.propTypes = {
  value: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  done: PropTypes.bool.isRequired,
};

export default TodoItem;
