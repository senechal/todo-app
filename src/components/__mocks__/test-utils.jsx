import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { render } from '@testing-library/react';
import reducer from '../../reducers';

export const renderWithRedux = (component, {
  initialState,
  store = createStore(reducer, initialState),
} = {}, options) => ({
  ...render(<Provider store={store}>{component}</Provider>, options),
  ...store,
});
