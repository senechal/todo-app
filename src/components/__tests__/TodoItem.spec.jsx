import React from 'react';
import 'jest-styled-components';
import { fireEvent } from '@testing-library/react';
import * as actions from 'actions';
import {
  TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT,
  TEST_ID_INLINE_EDIT_INPUT,
  TEST_ID_TODO_ITEM_CHECKBOX,
  TEST_ID_TODO_ITEM_DELETE_BUTTON,
  EMPTY_STRING,
} from 'utils/constants';
import TodoItem from '../TodoItem';
import { renderWithRedux } from '../__mocks__/test-utils';

jest.mock('lodash/delay', () => (fn) => { fn(); });

describe('Testing TodoItem', () => {
  let props;
  let initialState;
  beforeEach(() => {
    props = {
      value: 'value',
      id: 0,
      done: false,
    };
    initialState = {
      autoIncrement: 0,
      counter: 1,
      items: [props.id],
      details: {
        [props.id]: props,
      },
      editing: null,
    };
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should render correctly with required props', () => {
    const { container } = renderWithRedux(<TodoItem {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with changed value and id', () => {
    props.value = 'new value';
    props.id = 2;
    const { container } = renderWithRedux(<TodoItem {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with changed done state', () => {
    props.done = true;
    const { container } = renderWithRedux(<TodoItem {...props} />);
    expect(container).toMatchSnapshot();
  });

  it(`should change editing state when user start editing value and
  should change value on the store when user change input`, () => {
    const startEditingAction = jest.spyOn(actions, 'startEditing');
    const changeItemAction = jest.spyOn(actions, 'changeItem');
    const targetValue = 'change Value';
    const { getByTestId } = renderWithRedux(<TodoItem {...props} />);

    const itemContainer = getByTestId(TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT);
    fireEvent.click(itemContainer);
    const input = getByTestId(TEST_ID_INLINE_EDIT_INPUT);
    expect(startEditingAction).toBeCalledWith(props.id);
    fireEvent.change(input, { target: { value: targetValue } });

    expect(changeItemAction).toBeCalledWith(targetValue, props.id);
  });

  it('should toggle done when user click on delete button', () => {
    const removeItemAction = jest.spyOn(actions, 'removeItem');
    const { getByTestId } = renderWithRedux(<TodoItem {...props} />, { initialState });

    const button = getByTestId(TEST_ID_TODO_ITEM_DELETE_BUTTON);
    fireEvent.click(button);

    expect(removeItemAction).toBeCalledWith(props.id);
  });

  it('should delete item when user click on delete button', () => {
    const toggleItemAction = jest.spyOn(actions, 'toggleItem');
    const { getByTestId } = renderWithRedux(<TodoItem {...props} />, { initialState });

    const checkbox = getByTestId(TEST_ID_TODO_ITEM_CHECKBOX);
    fireEvent.click(checkbox);

    expect(toggleItemAction).toBeCalledWith(props.id);
  });

  it('should endEditing when user blur input', () => {
    const endEditingAction = jest.spyOn(actions, 'endEditing');
    const { getByTestId } = renderWithRedux(<TodoItem {...props} />);
    const itemContainer = getByTestId(TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT);
    fireEvent.click(itemContainer);

    const input = getByTestId(TEST_ID_INLINE_EDIT_INPUT);
    fireEvent.blur(input);
    expect(endEditingAction).toBeCalled();
  });

  it('should endEditing when user blur input and should call remove item if the value is empty', () => {
    const endEditingAction = jest.spyOn(actions, 'endEditing');
    const removeItemAction = jest.spyOn(actions, 'removeItem');
    props.value = EMPTY_STRING;
    const { getByTestId } = renderWithRedux(<TodoItem {...props} />, { initialState });
    const itemContainer = getByTestId(TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT);
    fireEvent.click(itemContainer);

    const input = getByTestId(TEST_ID_INLINE_EDIT_INPUT);
    fireEvent.blur(input);
    expect(endEditingAction).toBeCalled();
    expect(removeItemAction).toBeCalledWith(props.id);
  });
});
