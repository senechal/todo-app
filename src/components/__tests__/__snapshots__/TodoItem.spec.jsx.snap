// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TodoItem should render correctly with changed done state 1`] = `
.c0 {
  color: #4d4d4d;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 8px;
  height: 40px;
  border-bottom: 1px solid #e6e6e6;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
}

.c0:last-child:not(:nth-child(-n+2)) {
  border-bottom: 0px;
}

.c1 {
  margin-right: 8px;
}

.c4 {
  background-color: #fff;
  width: 24px;
  height: 24px;
  padding: 2.4px 0;
  font-size: 14px;
  vertical-align: -1px;
  border-radius: 4px;
  border: 1px solid #1890ff;
  cursor: pointer;
  margin-left: auto;
  background-color: #1890ff;
  color: #fff;
}

.c4:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c5 {
  background-color: #fff;
  border-color: #e6e6e6;
  color: #e6e6e6;
}

.c5 svg {
  fill: #e6e6e6;
}

.c3 {
  color: #4d4d4d font-size:16px;
  height: 20px;
  box-sizing: border-box;
  width: calc(100% - 62px);
}

.c3:hover {
  border-bottom: 1px solid #e6e6e6;
}

.c2 {
  color: #4d4d4d;
  outline: 0;
  border-width: 0 0 1px;
  border-color: #4d4d4d;
  font-size: 16px;
  height: 20px;
  box-sizing: border-box;
  display: none;
  width: calc(100% - 62px);
}

.c6 {
  max-width: 16px;
  max-height: 16px;
}

<div>
  <div
    class="c0"
  >
    <input
      checked=""
      class="c1"
      data-testid="todo-item-checkbox"
      type="checkbox"
    />
    <input
      class="c2"
      data-testid="inline-edit-input"
      type="text"
      value="value"
    />
    <div
      class="c3"
      data-testid="inline-edit-item-content"
    >
      value
    </div>
    <button
      class="c4 c5"
      data-testid="todo-item-delete-button"
      title="Delete"
    >
      <svg
        class="c6"
        data-file-name="SvgDelete"
      />
    </button>
  </div>
</div>
`;

exports[`Testing TodoItem should render correctly with changed value and id 1`] = `
.c0 {
  color: #4d4d4d;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 8px;
  height: 40px;
  border-bottom: 1px solid #e6e6e6;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
}

.c0:last-child:not(:nth-child(-n+2)) {
  border-bottom: 0px;
}

.c1 {
  margin-right: 8px;
}

.c4 {
  background-color: #fff;
  width: 24px;
  height: 24px;
  padding: 2.4px 0;
  font-size: 14px;
  vertical-align: -1px;
  border-radius: 4px;
  border: 1px solid #1890ff;
  cursor: pointer;
  margin-left: auto;
  background-color: #1890ff;
  color: #fff;
}

.c4:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c5 {
  background-color: #fff;
  border-color: #e6e6e6;
  color: #e6e6e6;
}

.c5 svg {
  fill: #e6e6e6;
}

.c3 {
  color: #4d4d4d font-size:16px;
  height: 20px;
  box-sizing: border-box;
  width: calc(100% - 62px);
}

.c3:hover {
  border-bottom: 1px solid #e6e6e6;
}

.c2 {
  color: #4d4d4d;
  outline: 0;
  border-width: 0 0 1px;
  border-color: #4d4d4d;
  font-size: 16px;
  height: 20px;
  box-sizing: border-box;
  display: none;
  width: calc(100% - 62px);
}

.c6 {
  max-width: 16px;
  max-height: 16px;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      data-testid="todo-item-checkbox"
      type="checkbox"
    />
    <input
      class="c2"
      data-testid="inline-edit-input"
      type="text"
      value="new value"
    />
    <div
      class="c3"
      data-testid="inline-edit-item-content"
    >
      new value
    </div>
    <button
      class="c4 c5"
      data-testid="todo-item-delete-button"
      title="Delete"
    >
      <svg
        class="c6"
        data-file-name="SvgDelete"
      />
    </button>
  </div>
</div>
`;

exports[`Testing TodoItem should render correctly with required props 1`] = `
.c0 {
  color: #4d4d4d;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 8px;
  height: 40px;
  border-bottom: 1px solid #e6e6e6;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
}

.c0:last-child:not(:nth-child(-n+2)) {
  border-bottom: 0px;
}

.c1 {
  margin-right: 8px;
}

.c4 {
  background-color: #fff;
  width: 24px;
  height: 24px;
  padding: 2.4px 0;
  font-size: 14px;
  vertical-align: -1px;
  border-radius: 4px;
  border: 1px solid #1890ff;
  cursor: pointer;
  margin-left: auto;
  background-color: #1890ff;
  color: #fff;
}

.c4:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c5 {
  background-color: #fff;
  border-color: #e6e6e6;
  color: #e6e6e6;
}

.c5 svg {
  fill: #e6e6e6;
}

.c3 {
  color: #4d4d4d font-size:16px;
  height: 20px;
  box-sizing: border-box;
  width: calc(100% - 62px);
}

.c3:hover {
  border-bottom: 1px solid #e6e6e6;
}

.c2 {
  color: #4d4d4d;
  outline: 0;
  border-width: 0 0 1px;
  border-color: #4d4d4d;
  font-size: 16px;
  height: 20px;
  box-sizing: border-box;
  display: none;
  width: calc(100% - 62px);
}

.c6 {
  max-width: 16px;
  max-height: 16px;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      data-testid="todo-item-checkbox"
      type="checkbox"
    />
    <input
      class="c2"
      data-testid="inline-edit-input"
      type="text"
      value="value"
    />
    <div
      class="c3"
      data-testid="inline-edit-item-content"
    >
      value
    </div>
    <button
      class="c4 c5"
      data-testid="todo-item-delete-button"
      title="Delete"
    >
      <svg
        class="c6"
        data-file-name="SvgDelete"
      />
    </button>
  </div>
</div>
`;
